Generate autograd functions for NN / more refactors

colesbury opened the pull request:
- Implement more derivatives in derivatives.yaml
- Handle sized IntList in python_arg_parser

colesbury committed:
Generate autograd functions for NN

Implement more derivatives in derivatives.yaml

colesbury commented:
Note that the Variable functions are still shadowed by the definitions in variable.py and the NN functions are not exposed in Python.  I have another commit that exposes the functions, but it will break onnx and the jit.

apaszke commented:
The JIT is broken because of C++ functions anyway. This is higher priority so I think we can start with exposing the new functions in Python, and then fix the JIT

colesbury committed:
Use _import_array instead of import_array

This provides more detailed error messages if something goes wrong and
has a consistent return type across Python versions.

colesbury commented:
`gen_variable_type.py` is quite large and does a lot of things. It would probably benefit from some clean up, but I'd like to wait until we have all the functionality.

fmassa commented on tools/autograd/derivatives.yaml:
```diff
@@ -129,73 +119,197 @@
-- name: gesv
+- name: gesv(Tensor self, Tensor A)
+  self: std::get<0>(gesv(grad, A.t()))
+  A: -at::mm(std::get<0>(gesv(grad, A.t())), solution.t())
```
nit: if we want to reuse intermediate computations (like the `gesv(grad, A.t())`), the recommended way will be to write the derivative in C++ and not rely on the yaml, right?

colesbury replied:
Yes and it's not currently supported for functions in derivatives.yaml. I'd like to write:

`derivatives.yaml`:
```yaml
name: gesv(Tensor self, Tensor A)
self, A: gesv_backward(grad, A, solution)
```

`tools/autograd/templates/Functions.cpp`:
```c++
std::tuple<Tensor, Tensor> gesv_backward(...) {
 ...
}
```

This will require some more modifications to `gen_variable_type.py`.

apaszke replied:
As you notices, `gen_variable_type` starts getting complicated, which is a good sign that we should stop extending the language of `derivatives.yaml`. I doubt that `gesv` is a significant bottleneck for models, so let's just keep it implemented in Python. That file should only be used as an easy method to port simple functions to C++

colesbury replied:
We're going to need it for other things too like prelu double backwards

colesbury replied:
We'll have in `derivatives.yaml`:

```yaml
- name: prelu_backward(Tensor grad_output, Tensor input, Tensor weight, std::array<bool, 2> output_mask):
  grad_output, input, weight: prelu_double_backward(grads, grad_output, input, weight)
```

In some C++ file (possibly Functions.cpp):

```c++
std::tuple<Tensor, Tensor, Tensor> prelu_double_backward(TensorList grads, const Tensor & grad_output, const Tensor & input, const Tensor & weight) {
 // implemenation from PReLUBackward.backward
}
```

(It's possible that the double backwards for `grad_output` could be "unfused" from the rest of the computation)

There's two pieces of functionality we need to support this:
1. Support multiple values on the lhs in derivatives.yaml (this should be a small change)
2. Support taking multiple "grad_outputs" in backwards functions. Currently, all our autogenerated backwards functions assume as single grad_output as input.

fmassa commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -8,6 +8,12 @@ using at::IntList;
+namespace {
+
+Tensor not_implemented(const char* name) {
+  throw std::runtime_error(std::string(name) + " is not differentiable");
```
I think this error message is a bit misleading, because the operations might be differentiable but the derivatives were just not yet implemented.

colesbury committed:
Change error message in 'not_implemented'

ezyang commented:
Hey @colesbury / @gchanan , yesterday I was running the autograd tests through the new ATen codepath and I noticed that broadcasting is not handled correctly by the new codepath. The problem is this isn't sophisticated enough:

```
- name: add(Tensor self, Tensor other, *, Scalar alpha=1)
  self: grad
  other: grad * alpha
```

The original code new to unexpand if necessary:

```
@staticmethod
def backward(ctx, grad_output):
        return maybe_unexpand(grad_output, ctx.a_size), maybe_unexpand_or_view(grad_output, ctx.b_size), None
```

What's our plan for handling this?

gchanan commented:
Is there a specific test case you are referring to?  When I last ran the tests the broadcasting tests they passed.

colesbury commented:
Broadcasting is handled in the forward pass *outside* of the differentiable function (s_add), so the backwards for `s_add` doesn't have to worry about it.

`add(a, b)` ends up calling (in pseudo-code):

```
if (a.size() != b.size()) {
 a = a.expand(<broadcast_size>)
 b = b.expand(<broadcast_size>)
}
return s_add(a, b);
```

The backwards for `expand` calls `reduce_to` which is the equivalent to `maybe_unexpand` in the old code.

See:
Type::add
VariableType::s_add
VariableType::expand
ExpandUtils.h

ezyang commented:
Here's the one that failed:

```
======================================================================
ERROR: test_basic_op_grad_fallback (__main__.TestAutograd)
Grad output might need to be reshaped to match the second argument.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pytorch/test/test_autograd.py", line 753, in test_basic_op_grad_fallback
    (x + y()).sum().backward()
  File "/data/users/ezyang/onnx-pytorch/pytorch/torch/autograd/variable.py", line 817, in __add__
    return self.add(other)
RuntimeError: The size of tensor a (6) must match the size of tensor b (2) at non-singleton dimension 1
```

I do have some other local changes (including deleting add).

colesbury commented:
That tests deprecated behavior: adding non-broadcastable tensors which have the same number of elements. Since we're not supporting that in ATen, we can delete the test.

colesbury commented:
In case it's helpful, this commit deletes in my fork deletes the tests of deprecated behavior and enables the methods on Variable and in torch.nn.functional:

https://github.com/colesbury/pytorch/commit/cef804368797ab2d94ab2fbc16c8a8ff357d1432

ezyang commented:
Thank you!

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -68,10 +70,18 @@
+
+DERIVATIVE_MULTI = CodeTemplate("""\
+if (should_compute_output({ ${idxs} })) {
+  auto output_mask = std::array<bool, ${n}>{
```
is this used anywhere?

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -138,10 +161,21 @@
+    'arange', 'eye', 'linspace', 'logspace', 'tensor', 'ones', 'ones_like',
+    'rand', 'randn', 'randperm', 'range', 'tensor', 'uniform', 'zeros',
+    'zeros_like', 'set_',
+    # these are only implemented on integral types
```
why only for them?

apaszke replied:
Uh nevermind. I didn't realize these are only logical ops (I though basic magic methods are there too)

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+
+def create_derivative(declaration, formula, output_indices, var_names):
+    returns = [r for r in declaration['returns'] if r.get('name') != 'self']
+    arguments = [arg for arg in declaration['arguments']]
```
`arguments = declaration['arguments']`?

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+    if any(arg['name'] == 'inplace' for arg in arguments):
+        for arg in arguments:
+            if arg['name'] == 'input':
+                returns += [arg]
```
this logic is a bit weird. If we find `inplace` in argument list, we look for `input` and add it to returned values? Is it empty before or what?

colesbury replied:
This is for NN functions. For something like: `relu(Tensor input, bool inplace)`. If inplace is true then `input` is also a return argument. It's important that if we save input for the backwards we do it in the `save_outputs` section not the `save_inputs` because it might be modified.

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+        'buffers': [] if buffers is None else buffers,
+        'saved_inputs': all_saved_variables(derivatives, 'saved_inputs'),
+        'saved_outputs': all_saved_variables(derivatives, 'saved_outputs'),
+    }
```
Eh I'm starting to hate that this data is so unstructured. We should use objects with well defined attributes in this file (sort of like `namedtuple`, but probably we want to keep it mutable)

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+        for declaration in declarations:
+            if declaration['name'] == name:
+                return declaration
+        return declarations[0]
```
uh what? how is an arbitrary (first) declaration canonical for a name?

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
-        for param in params:
-            if param == '' or param == '*':
+        for arg in canonical['arguments']:
+            if arg['name'] not in defn:
```
does this mean that we skipped some of ATen arguments? when is it valid to do that?

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+                num_inputs = ''
+                output_indices = '*'
+            else:
+                output_indices = [num_inputs]
```
What are `output_indices` and why are they sometimes a list and sometimes a string?

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+                output_indices = '*'
+            else:
+                output_indices = [num_inputs]
+                num_inputs += 1
```
same here. why do we set `num_inputs` to an empty string in the upper case?

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+    functions_by_name = defaultdict(list)
+    for func in autograd_functions:
+        functions_by_name[func['op']].append(func)
+    for op in sorted(functions_by_name.keys()):
```
is there any reason why it has to be sorted?

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
 
-        saved = []
-        for arg in option['python_arguments']:
+        fwd = declarations_by_name[name + '_forward'][0]
```
nit: `name + '_forward'` -> `forward_name`

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+        formula = '{}({})'.format(bwd_name, ', '.join(actuals))
+        formula = formula.replace('grad_output', 'grad')
+        if not re.search(IDENT_REGEX.format('grad'), formula):
+            formula = '({}).mul_(grad)'.format(formula)
```
Uhhh this makes me nervous. Maybe let's just raise an error if `grad` doesn't appear anywhere

apaszke replied:
Yes, but why can't we include that in the formula itself? Plus `mul_` with 2 tensors is going to raise an error in backward

colesbury replied:
I'm not sure what you mean. These four lines are constructing the formula (from the signature of the backwards function) for the first derivative of NN functions.  The formulas aren't specified anywhere else: it's implicit from the naming of xxx_forwards and xxx_backwards 

(The second derivatives are handled in derivatives.yaml).

apaszke replied:
Ohhhhh I see. Makes sense now. Thank you

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -309,74 +417,75 @@ def create_autograd_functions(top_env, declarations):
+                masks = ['should_compute_output({}),'.format(i) for i in idxs]
+                return DERIVATIVE_MULTI.substitute(
+                    idxs=idxs, derivative=formula, grad_inputs=grad_inputs,
+                    masks=masks, n=len(idxs))
```
can you quickly explain to me what masks and output_idxes are?

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -391,44 +500,105 @@ def create_variable_type(top_env, aten_declarations):
+                if option['inplace'] and is_output:
+                    var = 'self'
+                ptr = 'grad_fn.get()' if is_output else 'nullptr'
+                expr = 'SavedVariable({}, {})'.format(var, ptr)
```
can you add an `else` that fails?

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -502,17 +686,28 @@ def load_aten_declarations(path):
             if arg.get('default') is not None:
                 default = arg['default']
-                if default == 'nullptr':
+                if default == 'nullptr' or default == '{}':
```
what's a `{}` default? Is it an uninitialized Tensor?

apaszke commented on torch/csrc/autograd/function.h:
```diff
@@ -121,6 +121,15 @@ struct Function : std::enable_shared_from_this<Function> {
   }
 
+  inline bool should_compute_output(std::initializer_list<int> idxs) const {
+    for (auto idx : idxs) {
```
nit:
```cpp
return std::any_of(idxs.begin(), idxs.end(), should_compute_outputs);
```

apaszke reviewed (approved):
LGTM. I only have a lot of questions and very minor comments

colesbury committed:
changes from review

colesbury committed:
Use masked_fill in min/max backwards

colesbury committed:
Fixes after rebasing:

 - Move generated functions to torch::autograd::generated namespace to
   avoid name conflicts
 - Add dim to [log_]softmax derivatives

colesbury committed:
Additional changes from review

colesbury commented on test/test_nn.py:
```diff
@@ -3507,7 +3507,8 @@ def add_test(test):
     setattr(TestNN, test_name, lambda self, test=test: test(self))
-    setattr(TestNN, cuda_test_name, lambda self, test=test: test.test_cuda(self))
+    if test_name != 'test_Hardshrink':  # no CUDA implementation
+        setattr(TestNN, cuda_test_name, lambda self, test=test: test.test_cuda(self))
```
Hmmm.. this change doesn't seem necessary any more. Removed.

apaszke commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -25,20 +32,45 @@ Tensor maybe_multiply(const Tensor & t, const Scalar & s) {
   } else {
     auto pow_ = self.abs().pow(p - 2);
-    auto scale_v = grad / self.norm(p).toTensor().pow(p - 1);
+    auto scale_v = grad / norm.toTensor().pow(p - 1);
```
I thought of using the C `pow` from stdlib

colesbury replied:
Hmmm... let met try this out. Also I think we can save the scalar output (although it get's converted to a 1D Variable via `as_variable`).

colesbury replied:
Saving the result helps, but we don't want to use `std::pow` because then we'd break double-backwards for this function.  (Tensor * Scalar would have the same problem; currently (no gradient for the Scalar).

We'll want to fix these as we get true 0-dim tensor support. (It's not an issue now because Scalars are never exposed to Python; we wrap them in 1-dim Variables).

colesbury replied:
See #3188
